src/catalog/src/durable/impls/persist/state_update.rs

```diff
@@ -7,29 +7,55 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use prost::Message;
 use std::fmt::Debug;
 
 use mz_persist_types::codec_impls::{SimpleDecoder, SimpleEncoder, SimpleSchema};
 use mz_persist_types::dyn_struct::{ColumnsMut, ColumnsRef, DynStructCfg};
 use mz_persist_types::Codec;
 use mz_proto::{ProtoType, RustType, TryFromProtoError};
 use mz_repr::Diff;
+use proptest_derive::Arbitrary;
+use prost::Message;
 
 use crate::durable::impls::persist::Timestamp;
 use crate::durable::objects::serialization::proto;
 use crate::durable::transaction::TransactionBatch;
 use crate::durable::Epoch;
 
+/// Trait for objects that can be converted to/from a [`StateUpdateKindBinary`].
+pub(crate) trait IntoStateUpdateKindBinary:
+    Into<StateUpdateKindBinary>
+    + TryFrom<StateUpdateKindBinary>
+    + PartialEq
+    + Eq
+    + PartialOrd
+    + Ord
+    + Debug
+    + Clone
+{
+}
```

Comment:
The trait approach works, and I think is the right one. An idea that popped into my head was using an enum instead, e.g.

```
enum StateUpdateKindVersioned {
  // ... all the other versions
  V45(...),
  Current(...),
}
```

as I typed that out I like it less, just figured I would run it by you though. I still think the trait works best here though because we for the Catalog we can statically guarantee we're using the latest protos

Reply:
I think it's a good idea, but I also like the trait approach slightly better. It also has the benefit of being automatically implemented by all proto versions.